File: Homework5/assignment5.py
# -*- coding: utf-8 -*-
# SVR

# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
dataset = pd.read_csv('Housing_Data.csv')
X = dataset.iloc[:, 0:-1].to_numpy()
y = dataset.iloc[:, -1].to_numpy()

#Splitting dataset into training and test set
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25)

# Feature Scaling
from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
sc_y = StandardScaler()
X_scaled = sc_X.fit_transform(X_train)
y_scaled = sc_y.fit_transform(y_train.reshape(len(y_train), 1)).flatten()
X_test_scaled = sc_X.fit_transform(X_test)

# Training the SVR model on the whole dataset
from sklearn.svm import SVR
regressor_svr = SVR(kernel = 'rbf')  
regressor_svr.fit(X_scaled, y_scaled)
y_pred_scaled = regressor_svr.predict(X_test_scaled) #predict y values using x-test-scaled values
y_pred = sc_y.inverse_transform(y_pred_scaled.reshape(len(y_pred_scaled), 1)) #Reverse the scaling


from sklearn.metrics import r2_score
r2 = r2_score(y_test, y_pred)
print(r2)
r2_adjusted = 1 - (((1 - r2) * (len(y_test) - 1))/ (len(y_test) - 6 - 1))
print(r2_adjusted)

#.69 1
#.68 2
#.53 3
#.80 4
#.75 5
#.66 6
#.71 7
#.66 8
#.56 9
#.74 10
#avg = .678


# Random Forest Regression

# Importing the dataset
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
dataset = pd.read_csv('Housing_Data.csv')
X = dataset.iloc[:, 0:-1].to_numpy()
y = dataset.iloc[:, -1].to_numpy()

from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25) #Split data into training and test

# Training the Random Forest Regression model on the whole dataset
from sklearn.ensemble import RandomForestRegressor
regressor_forest = RandomForestRegressor(n_estimators = 500)
regressor_forest.fit(X_train, y_train)
y_pred_forest = regressor_forest.predict(X_test)

from sklearn.metrics import r2_score

r2 = r2_score(y_test, y_pred_forest)
print(r2)
r2_adjusted = (1 - (((1 - r2) * (len(y_test) - 1))/ (len(y_test) - 6 - 1)))
print(r2_adjusted)
#.78 1
#.76 2
#.55 3
#.81 4
#.73 5
#.61 6
#.76 7
#.73 8
#.57 9
#.65 10
# avg = .695

#The forest tree regression is slightly better with an average of .695 adjusted r-squared

File: Homework5/assignment5.R
dataset = read.csv('Housing_data.csv')


# Splitting the dataset into the Training set and Test set
library(caTools)
split = sample.split(dataset$Y, SplitRatio = 3/4)

training_set = subset(dataset, split == TRUE)
test_set = subset(dataset, split == FALSE)

# Fitting SVR to the whole dataset
library(e1071)
regressor = svm(formula = Y ~ .,
                data = training_set,
                type = 'eps-regression',
                kernel = 'radial')
y_pred = predict(regressor, test_set)

#Find R^2 and adjusted R^2
ssr = sum((test_set$Y - y_pred) ^ 2)
sst = sum((test_set$Y - mean(test_set$Y)) ^ 2)
r2 = 1 - (ssr/sst)
print(r2)
r2_adjusted = 1 - (1 - r2) * (length(test_set$Y) - 1) / (length(test_set$Y) - 6 - 1)
print(r2_adjusted) 

#.59
#.54
#.61
#.70
#.76
#.68
#.73
#.72
#.61
#.67
#avg = 0.656


# Random Forest Regression


#Splitting data in training and test set
library(caTools)
split = sample.split(dataset$Y, SplitRatio = 3/4)

training_set = subset(dataset, split == TRUE)
test_set = subset(dataset, split == FALSE)

# Fitting Random Forest Regression to the whole dataset
library(randomForest)
regressor = randomForest(formula = Y ~ .,
                         data = training_set,
                         ntree = 500)

y_pred = predict(regressor, newdata = test_set)

#Calculate R^2 and adjusted R^2
ssr = sum((test_set$Y - y_pred) ^ 2)
sst = sum((test_set$Y - mean(test_set$Y)) ^ 2)
r2 = 1 - (ssr/sst)
print(r2)
r2_adjusted = 1 - (1 - r2) * (length(test_set$Y) - 1) / (length(test_set$Y) - 6 - 1)
print(r2_adjusted) 

#.57
#.71
#.69
#.75
#.78
#.64
#.76
#.83
#.76
#.74
#avg = 0.723

#The random forest tree is the better model with an avg of .723